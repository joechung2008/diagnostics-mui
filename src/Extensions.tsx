import { Button, Grid, List, ListItem } from "@mui/material";
import { byKey, isExtensionInfo, toNavLink } from "./utils";

const Extensions: React.FC<ExtensionsProps> = ({ extensions, onLinkClick }) => {
  const links = Object.values(extensions)
    .filter(isExtensionInfo)
    .map(toNavLink)
    .sort(byKey);

  return (
    <Grid className="extension-root" size="auto">
      <List aria-label="Extensions">
        {links.map((link) => (
          <ListItem key={link.key}>
            <Button
              className="nav-button"
              sx={{ justifyContent: "flex-start" }}
              onClick={(e) => onLinkClick?.(e, link)}
            >
              {link.name}
            </Button>
          </ListItem>
        ))}
      </List>
    </Grid>
  );
};

export default Extensions;
