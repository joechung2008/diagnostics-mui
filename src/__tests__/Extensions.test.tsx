import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import Extensions from "../Extensions";

describe("Extensions", () => {
  it("matches snapshot with minimal extension", () => {
    const extensions = {
      ext1: { extensionName: "ext1" },
    };
    const onLinkClick = () => {};
    const { container } = render(
      <Extensions extensions={extensions} onLinkClick={onLinkClick} />
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with multiple extensions", () => {
    const extensions = {
      ext1: { extensionName: "ext1" },
      ext2: { extensionName: "ext2" },
    };
    const onLinkClick = () => {};
    const { container } = render(
      <Extensions extensions={extensions} onLinkClick={onLinkClick} />
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with config", () => {
    const extensions = {
      ext1: {
        extensionName: "ext1",
        config: { foo: "bar" },
      },
    };
    const onLinkClick = () => {};
    const { container } = render(
      <Extensions extensions={extensions} onLinkClick={onLinkClick} />
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with stageDefinition", () => {
    const extensions = {
      ext1: {
        extensionName: "ext1",
        stageDefinition: { name: ["Stage1"] },
      },
    };
    const onLinkClick = () => {};
    const { container } = render(
      <Extensions extensions={extensions} onLinkClick={onLinkClick} />
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with config and stageDefinition", () => {
    const extensions = {
      ext1: {
        extensionName: "ext1",
        config: { foo: "bar" },
        stageDefinition: { name: ["Stage1"] },
      },
    };
    const onLinkClick = () => {};
    const { container } = render(
      <Extensions extensions={extensions} onLinkClick={onLinkClick} />
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with ExtensionError", () => {
    const extensions = {
      ext1: {
        extensionName: "ext1",
      },
      ext2: {
        lastError: {
          errorMessage: "Something went wrong",
          time: "2025-08-10T12:00:00Z",
        },
      },
    };
    const onLinkClick = () => {};
    const { container } = render(
      <Extensions extensions={extensions} onLinkClick={onLinkClick} />
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with no extensions", () => {
    const extensions = {};
    const onLinkClick = () => {};
    const { container } = render(
      <Extensions extensions={extensions} onLinkClick={onLinkClick} />
    );
    expect(container).toMatchSnapshot();
  });

  // Removed test for undefined extensions as the component does not handle undefined gracefully.

  it("calls onLinkClick when a link is clicked", () => {
    const extensions = {
      ext1: { extensionName: "ext1" },
      ext2: { extensionName: "ext2" },
    };
    const onLinkClick = vi.fn();
    const { container } = render(
      <Extensions extensions={extensions} onLinkClick={onLinkClick} />
    );
    const buttons = screen.getAllByRole("button");
    buttons.forEach((button) => fireEvent.click(button));
    expect(onLinkClick).toHaveBeenCalledTimes(2);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with mixed ExtensionInfo and ExtensionError", () => {
    const extensions = {
      ext1: { extensionName: "ext1" },
      ext2: {
        lastError: {
          errorMessage: "Error for ext2",
          time: "2025-08-10T13:00:00Z",
        },
      },
    };
    const onLinkClick = () => {};
    const { container } = render(
      <Extensions extensions={extensions} onLinkClick={onLinkClick} />
    );
    expect(container).toMatchSnapshot();
  });
});
