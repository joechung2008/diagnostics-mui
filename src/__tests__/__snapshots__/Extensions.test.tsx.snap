// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Extensions > calls onLinkClick when a link is clicked 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
  >
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            ext1
          </span>
        </div>
      </div>
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            ext2
          </span>
        </div>
      </div>
    </ul>
  </div>
</div>
`;

exports[`Extensions > matches snapshot with ExtensionError 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
  >
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            ext1
          </span>
        </div>
      </div>
    </ul>
  </div>
</div>
`;

exports[`Extensions > matches snapshot with config 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
  >
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            ext1
          </span>
        </div>
      </div>
    </ul>
  </div>
</div>
`;

exports[`Extensions > matches snapshot with config and stageDefinition 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
  >
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            ext1
          </span>
        </div>
      </div>
    </ul>
  </div>
</div>
`;

exports[`Extensions > matches snapshot with minimal extension 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
  >
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            ext1
          </span>
        </div>
      </div>
    </ul>
  </div>
</div>
`;

exports[`Extensions > matches snapshot with mixed ExtensionInfo and ExtensionError 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
  >
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            ext1
          </span>
        </div>
      </div>
    </ul>
  </div>
</div>
`;

exports[`Extensions > matches snapshot with multiple extensions 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
  >
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            ext1
          </span>
        </div>
      </div>
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            ext2
          </span>
        </div>
      </div>
    </ul>
  </div>
</div>
`;

exports[`Extensions > matches snapshot with no extensions 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
  >
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
    />
  </div>
</div>
`;

exports[`Extensions > matches snapshot with stageDefinition 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-auto extension-root css-1hef13s-MuiGrid-root"
  >
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            ext1
          </span>
        </div>
      </div>
    </ul>
  </div>
</div>
`;
